smaps/src/main/java/com/google/smaps/FileUpload.java

```diff
@@ -16,61 +16,80 @@
+ * Retrieves the uploaded file from the user and then saves it to the tmp directory, while also
+ * setting attributes to this user's session.
  */
 @MultipartConfig
 @WebServlet(name = "FileUpload", value = "/fileupload")
 public class FileUpload extends HttpServlet {
-  // This String is sent back to index.html with a message if there's an error with file upload.
-  String errorMessage = "";
-
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response)
       throws IOException, ServletException {
-    // Resets the error message.
-    errorMessage = "";
+    // Get the current session, which contains user-specific data.
+    HttpSession session = request.getSession();
+
+    // The name of this user's file upload, will later be set to a file path.
+    String filename;
 
-    // Check whether example mode was selected, if it wasn't selected the string will be null.
+    // Reset the error message to be empty, it will later be set to a message if there's an error
+    // with upload.
+    String fileErrorMessage = "";
+    session.setAttribute("fileErrorMessage", fileErrorMessage);
+
+    // Check whether example mode was selected, if it wasn't selected the exampleButton will be
+    // null.
     String exampleButton = request.getParameter("exampleMode");
     if (exampleButton != null) {
-      // Create an example input stream from the smaps-full file, and upload it.
-      File exampleFile = new File("../smaps-full.txt");
-      InputStream exampleInputStream = new FileInputStream(exampleFile);
-      uploadFile(exampleInputStream);
+      // Create an input stream from the example smaps file.
+      InputStream exampleInputStream = new FileInputStream(new File("WEB-INF/smaps-example.txt"));
+
+      // Get the randomized filename from uploadFile and set the filename to the session.
+      filename = uploadFile(session, exampleInputStream);
+      session.setAttribute("filename", filename);
 
       // Send user to the histogram page.
       response.sendRedirect("/interactive-histogram.html");
+
     } else {
-      // Gets the file chosen by the user.
+      // Get the file chosen by the user.
       Part filePart = request.getPart("smapsFile");
 
-      // TODO(@sophbohr): add more error cases with printouts.
-
-      // Checks if the file part is empty (meaning no file was selected or the file was empty).
+      // TODO(@sophbohr): Add more error checking.
+      // Check if the file part is empty (meaning no file was selected or the file was empty).
       if (filePart.getSize() == 0) {
-        errorMessage = "Error: No file chosen or file was empty.";
+        // Set the error message.
+        fileErrorMessage = "Error: No file chosen or file was empty.";
+        session.setAttribute("fileErrorMessage", fileErrorMessage);
         // Send user back to the index.html page with error message printed with doGet.
         response.sendRedirect("/index.html");
       } else {
```

Comment:
Can you return at the end of the if here and remove the else?

Reply:
Yes!